Signal/src/ViewControllers/ContactViewController.swift

```diff
@@ -0,0 +1,686 @@
+            stackView.autoPinLeadingToSuperviewMargin(withInset: hMargin)
+            stackView.autoPinTrailingToSuperviewMargin(withInset: hMargin)
+            lastView = stackView
+        case .systemContactWithoutSignal:
+            // Show invite button for system contacts without a Signal account.
+            let inviteButton = createLargePillButton(text: NSLocalizedString("ACTION_INVITE",
+                                                                                          comment: "Label for 'invite' button in contact view."),
+                                                     actionBlock: { [weak self] _ in
+                                                        guard let strongSelf = self else { return }
+                                                        strongSelf.didPressInvite()
+            })
+            topView.addSubview(inviteButton)
+            inviteButton.autoPinEdge(.top, to: .bottom, of: lastView, withOffset: 20)
+            inviteButton.autoPinLeadingToSuperviewMargin(withInset: 55)
+            inviteButton.autoPinTrailingToSuperviewMargin(withInset: 55)
+            lastView = inviteButton
+        case .nonSystemContact:
+            // Show no action buttons for contacts not in user's device contacts.
+            break
+        case .noPhoneNumber:
+            // Show no action buttons for contacts without a phone number.
+            break
+        case .unknown:
+            let activityIndicator = UIActivityIndicatorView(activityIndicatorStyle: .whiteLarge)
+            topView.addSubview(activityIndicator)
+            activityIndicator.autoPinEdge(.top, to: .bottom, of: lastView, withOffset: 10)
+            activityIndicator.autoHCenterInSuperview()
+            lastView = activityIndicator
+            break
+        }
+
+        lastView.autoPinEdge(toSuperviewEdge: .bottom, withInset: 15)
+
+        let bottomView = UIView.container()
+        bottomView.backgroundColor = UIColor.white
+        bottomView.layoutMargins = .zero
+        bottomView.preservesSuperviewLayoutMargins = false
+        rootView.addSubview(bottomView)
+        bottomView.autoPinEdge(.top, to: .bottom, of: topView, withOffset: 0)
+        bottomView.autoPinEdge(toSuperviewEdge: .bottom)
+        bottomView.autoPinEdge(.left, to: .left, of: self.view, withOffset: 0)
+        bottomView.autoPinEdge(.right, to: .right, of: self.view, withOffset: 0)
+        bottomView.setContentHuggingVerticalLow()
+
+        var lastRow: UIView?
+
+        let addSpacerRow = {
+            guard let prevRow = lastRow else {
+                owsFail("\(self.logTag) missing last row")
+                return
+            }
+            let row = UIView()
+            row.backgroundColor = UIColor(rgbHex: 0xdedee1)
+            bottomView.addSubview(row)
+            row.autoSetDimension(.height, toSize: 1)
+            row.autoPinLeadingToSuperviewMargin(withInset: self.hMargin)
```

Comment:
No biggie, but since we're using this `hMargin` all over the place, did you consider defining an actual layout margin with edge insets and pinning directly to them rather than explicitly setting the offset for all these views? (and pinning to the Edge when you don't want the margin)

Reply:
Yup, I've been following your lead of trying to use layout margins more, but I haven't been strict about always going back and making sure we use them in every possible place.